### ISSUE TITLE
Find a refCnt issue in HttpPostRequestEncoder.java when I run HttpUploadClient.java example on netty-netty-4.0.25.Final

### ISSUE THREAD

**Issue by junfangwang** [netty/netty#3308]
# Issue:
When I run HttpUploadClient.java example on netty-netty-4.0.25.Final, it works well.
But when I make a little change on textAreaLong parameter’s value which was sent in HttpUploadClient.java as below:
change from:

            bodyRequestEncoder.addBodyAttribute("fourthinfo", textAreaLong); 

to:

            bodyRequestEncoder.addBodyAttribute("fourthinfo", textAreaLong.substring(0, 7470));

I met below issue and the request couldn’t be sent successfully.

```
12:56:57.651 [nioEventLoopGroup-2-3] WARN  i.n.channel.DefaultChannelPipeline - Failed to fail the promise because it's done already: DefaultChannelPromise@5324e236(failure(io.netty.util.IllegalReferenceCountException: refCnt: 0)
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:101) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.AbstractDerivedByteBuf.release(AbstractDerivedByteBuf.java:50) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.CompositeByteBuf$Component.freeIfNecessary(CompositeByteBuf.java:1369) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.CompositeByteBuf.deallocate(CompositeByteBuf.java:1624) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:106) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.AbstractDerivedByteBuf.release(AbstractDerivedByteBuf.java:50) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.CompositeByteBuf$Component.freeIfNecessary(CompositeByteBuf.java:1369) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.CompositeByteBuf.deallocate(CompositeByteBuf.java:1624) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:106) ~[netty-buffer-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:59) ~[netty-common-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.write(AbstractChannel.java:670) ~[netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.write(DefaultChannelPipeline.java:1113) ~[netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:633) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:691) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:626) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:125) [netty-codec-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:192) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:633) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:691) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:626) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:612) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:260) [netty-handler-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:139) [netty-handler-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:663) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext.access$1500(AbstractChannelHandlerContext.java:32) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.AbstractChannelHandlerContext$16.run(AbstractChannelHandlerContext.java:651) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:380) [netty-common-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357) [netty-transport-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116) [netty-common-4.0.25.Final.jar:4.0.25.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [netty-common-4.0.25.Final.jar:4.0.25.Final]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
```

# Fixed Issue:
After debug the code, I found that this issue is because when I change the length of parameter’s value, it will call fillByteBuf method in HttpPostRequestEncoder.java, which there is a refCnt issue here.
When I change the code on fillByteBuf method in HttpPostRequestEncoder.java from:

            ByteBuf slice = currentBuffer.slice(currentBuffer.readerIndex(), HttpPostBodyUtil.chunkSize);

to:

            ByteBuf slice = currentBuffer.slice(currentBuffer.readerIndex(), HttpPostBodyUtil.chunkSize);
            currentBuffer.retain();

The above issue was fixed and works well.

**Comment by fredericBregier** [comment]
@junfangwang Thank you for reporting this issue.
I recheck on my side and I confirm the issue. I submit a PR for that (see #3309).

**Comment by normanmaurer** [comment]
@junfangwang thanks.. fixed by #3309
